src/modules/land_detector/MulticopterLandDetector.cpp

```diff
@@ -154,40 +149,88 @@ bool MulticopterLandDetector::_get_ground_contact_state()
 		return true;
 	}
 
+	const bool lpos_available = (hrt_elapsed_time(&_vehicle_local_position.timestamp) < 1_s);
+
 	// land speed threshold
-	float land_speed_threshold = 0.9f * math::max(_params.landSpeed, 0.1f);
+	const float land_speed_threshold = 0.9f * math::max(_params.landSpeed, 0.1f);
+
 
-	// Check if we are moving vertically - this might see a spike after arming due to
-	// throttle-up vibration. If accelerating fast the throttle thresholds will still give
-	// an accurate in-air indication.
 	bool vertical_movement = false;
 
-	if (hrt_elapsed_time(&_landed_time) < LAND_DETECTOR_LAND_PHASE_TIME_US) {
+	if (lpos_available && _vehicle_local_position.v_z_valid) {
+		// Check if we are moving vertically - this might see a spike after arming due to
+		// throttle-up vibration. If accelerating fast the throttle thresholds will still give
+		// an accurate in-air indication.
+		float max_climb_rate = math::min(land_speed_threshold * 0.5f, _param_lndmc_z_vel_max.get());
 
-		// Widen acceptance thresholds for landed state right after arming
-		// so that motor spool-up and other effects do not trigger false negatives.
-		vertical_movement = fabsf(_vehicle_local_position.vz) > _param_lndmc_z_vel_max.get()  * 2.5f;
+		if (hrt_elapsed_time(&_landed_time) < LAND_DETECTOR_LAND_PHASE_TIME_US) {
+			// Widen acceptance thresholds for landed state right after arming
+			// so that motor spool-up and other effects do not trigger false negatives.
+			max_climb_rate = _param_lndmc_z_vel_max.get() * 2.5f;
+		}
 
-	} else {
-		// Adjust max_climb_rate if land_speed is lower than 2x max_climb_rate
-		float max_climb_rate = ((land_speed_threshold * 0.5f) < _param_lndmc_z_vel_max.get()) ? (0.5f * land_speed_threshold) :
-				       _param_lndmc_z_vel_max.get();
-		vertical_movement = fabsf(_vehicle_local_position.vz) > max_climb_rate;
+		vertical_movement = (fabsf(_vehicle_local_position.vz) > max_climb_rate);
 	}
 
+
 	// Check if we are moving horizontally.
-	_horizontal_movement = sqrtf(_vehicle_local_position.vx * _vehicle_local_position.vx
-				     + _vehicle_local_position.vy * _vehicle_local_position.vy) > _param_lndmc_xy_vel_max.get();
+	if (lpos_available && _vehicle_local_position.v_xy_valid) {
+
+		float gndspeed = sqrtf(_vehicle_local_position.vx * _vehicle_local_position.vx
+				       + _vehicle_local_position.vy * _vehicle_local_position.vy);
+
+		_horizontal_movement = (gndspeed > _param_lndmc_xy_vel_max.get());
+
+	} else {
+		_horizontal_movement = false;
```

Comment:
I don't think that if there is no local position, assuming that there is no horizontal movement as default is good default.
anyway, is there really reason not to have local position? or it just 'good practice'?

Reply:
Yes, when things go wrong.

Reply:
@dagar, so the comment is even more important, if there is no local position, we should no let the drone assume no _horizontal_movement which might lead to ground contact.

Reply:
Currently the final logic is `return ground_contact && !_horizontal_movement && !vertical_movement;`, so the default is set to remove `_horizontal_movement` from consideration.

Then consider https://github.com/PX4/Firmware/pull/15083/files#r458150562, where ground_contact isn't really used at all if there's no local position.

All that is to say, I think the horizontal_movement default like this makes sense, but there should still be additional logic for the case where there's no local position (although it doesn't currently matter either way).